### PR TITLE
feat: make agent starting period configurable [DET-3219]

### PR DESCRIPTION
**Pull request by shiyuann**
## Description

The provisioner would retry launching an agent after reaching
a duration, which is set by max agent starting period. Previously,
this configuration is hardcoded to 300 seconds. Now, change it to
be in the configuration of the provisioner.

## Test Plan

1. Manually test.

The goal of the test is to test if the functionality has been changed and the agent starting period is configurable as planned.

Use the following provisioner config:
```
provisioner:
  agent_docker_network: host
  agent_docker_image: determinedai/determined-dev:determined-agent-d90294d941964f104b50204232002fa551daa648
  max_idle_agent_period: 1m
  max_agent_starting_period: 10s
  operation_timeout_period: 1m
  provider: gcp
  boot_disk_size: 200
  boot_disk_source_image: projects/determined-ai/global/images/pedl-environments-4e98289
  network_interface:
    network: projects/determined-ai/global/networks/default
    subnetwork: projects/determined-ai/regions/us-central1/subnetworks/default
    external_ip: false
  instance_type:
    machine_type: n1-standard-8
    gpu_type: nvidia-tesla-v100
    gpu_num: 2
    preemptible: false
  max_instances: 2
```

2. Add unit tests for the refactoring.

## Commentary (optional)

Also rewrite some code to add more readability without changing the functionality.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: master/internal/provisioner/scale_decider.go [hunk 1]
```diff
@@ -240,3 +204,22 @@ func min(a, b int) int {
 	}
 	return b
 }
+
+func findLongLived(
+	candidate map[string]time.Time, stillLived map[string]bool, duration time.Duration,
```
Comment (aaron276h) [inline master/internal/provisioner/scale_decider.go:209]: non-blocking: rename `stilllived` to `idleInstances`

File: master/internal/provisioner/scale_decider.go [hunk 2]
```diff
@@ -240,3 +204,22 @@ func min(a, b int) int {
 	}
 	return b
 }
+
+func findLongLived(
```
Comment (aaron276h) [inline master/internal/provisioner/scale_decider.go:208]: ?: `funcLongLived` -> `funcLongIdle`

File: master/internal/provisioner/scale_decider.go [hunk 3]
```diff
@@ -98,78 +100,44 @@ func (s *scaleDecider) updateInstanceSnapshot(instances []*Instance) bool {
...
-		case ok && now.After(t.Add(s.maxIdleAgentPeriod)):
-			toTerminate[instID] = true
-		case ok:
-			instancesMarked[instID] = t
-		default:
-			instancesMarked[instID] = now
-		}
+	// Terminate instances that are idle for a long time.
+	var longLived map[string]bool
+	s.idleInstances, longLived = findLongLived(s.idleInstances, idleInstances, s.maxIdlePeriod)
```
Comment (aaron276h) [inline master/internal/provisioner/scale_decider.go:121]: ?: `longLived`-> `longIdle` (don't love this name either but is more accurate imo).

File: master/internal/provisioner/scale_decider.go [hunk 4]
```diff
@@ -7,87 +7,89 @@ import (
...
 	return &scaleDecider{
-		maxAgentStartingPeriod: maxAgentStartingPeriod,
-		maxIdleAgentPeriod:     maxIdleAgentPeriod,
-		schedulerSnapshot:      &scheduler.ViewSnapshot{},
+		maxStartingPeriod: maxStartingPeriod,
+		maxIdlePeriod:     maxIdlePeriod,
 	}
 }
 
+// needScale returns if a cluster needs rescaling.
```
[the remark this answers is not in the document]
Comment (shiyuann) [reply master/internal/provisioner/scale_decider.go:48]: considering it returns after the time pass over the minimum interval, yes.

File: master/internal/provisioner/scale_decider.go (hunk 1 above)
Comment (shiyuann) [reply master/internal/provisioner/scale_decider.go]: Talked offline: we decide to use `pastIdleInstances` and `presentIdleInstances`

**Pull request merged**